data-generation/generate_instructions_test.py

```diff
@@ -0,0 +1,275 @@
+
+class TestScatter:
+    def setup_method(self):
+        self.unscattered_blocks = [gi.Block('A', 'BLUE', 'B', 'GREEN', (1, 1), 1),
+                                   gi.Block('A', 'BLUE', 'B', 'GREEN', (1, 1), 2),
+                                   gi.Block('A', 'BLUE', 'B', 'GREEN', (1, 1), 3),
+                                   gi.Block('A', 'BLUE', 'B', 'GREEN', (1, 1), 4)]
+
+        self.scattered_blocks = [gi.Block('A', 'BLUE', 'B', 'GREEN', (0, 0), 1),
+                                 gi.Block('A', 'BLUE', 'B', 'GREEN', (0, 0), 2),
+                                 gi.Block('A', 'BLUE', 'B', 'GREEN', (0, 0), 3),
+                                 gi.Block('A', 'BLUE', 'B', 'GREEN', (0, 0), 4)]
+        self.randomizer_index = 0
+
+        def mock_randomize_block(block):
+            to_return = self.scattered_blocks[self.randomizer_index]
+            self.randomizer_index += 1
+            return to_return
+
+        self.orig_randomize_block = gi.randomize_block
+        gi.randomize_block = mock_randomize_block
+
+    def teardown_method(self):
+        gi.randomize_block = self.orig_randomize_block
+
+    def test_scatter_empty(self):
+        config = gi.Configuration([])
+        scattered_config = config.scatter()
+
+        assert scattered_config.current_blocks == []
+        assert scattered_config.final_blocks == []
+
+    def test_scatter_only_current(self):
+        base_config = gi.Configuration(self.unscattered_blocks[:])
+        scattered_config = base_config.scatter()
+
+        assert scattered_config.final_blocks == []
+        assert scattered_config.current_blocks == self.scattered_blocks
+
+    def test_scatter_current_and_final(self):
+        base_config = gi.Configuration(self.unscattered_blocks[0:2],
+                                       self.unscattered_blocks[2:4])
+        scattered_config = base_config.scatter()
+
+        assert scattered_config.final_blocks == []
+        assert scattered_config.current_blocks == self.scattered_blocks
+
+def test_random_block():
+    indexer = {}
+    indexer['rand_index'] = 0
+    def mock_rand_element(ls):
+        to_return = ls[indexer['rand_index']]
+        indexer['rand_index'] += 1
+        return to_return
+
+    orig_rand_element = gi.rand_element
+    gi.rand_element = mock_rand_element
+
+    def mock_random_position():
+        return random_position
+
+    orig_random_position = gi.random_position
+    gi.random_position = mock_random_position
+
+    random_position = (23, 54)
+
+    letters = ['A', 'B', 'C', 'D']
+    colors = ['BLUE', 'GREEN', 'RED', 'YELLOW']
+    block = gi.random_block(letters, colors)
+
+    gi.rand_element = orig_rand_element
+    gi.random_position = orig_random_position
+
+    assert block.side1_letter in letters
+    assert block.side2_letter in letters
+    assert block.side1_color in colors
+    assert block.side2_color in colors
+    assert block.position == random_position
```

Comment:
Why mock the random functions here? The letter and color assertions should hold with or without randomness, and I would think that checking that the position is a tuple containing two numbers would be enough.